docs: svc-mesh: Create some documentation about service mesh

sboeuf opened the pull request:
This commit introduces the intructions to be able to run trendy
service mesh Istio and Conduit with Kata Containers. It provides
a bit of feedback on how they actually work to give the reader a
quick overview. After this introduction, it provides restrictions
and instructions to enable them with Kata Containers.

Fixes #171

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
/cc @egernst @jodh-intel

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+in the control plane, and inject a proxy as a sidecar inside the service
+pod. The proxy registers to the control plane as a first step, and it
+constantly sends different sort of information about the service running
+inside the pod. Those information come from the filtering performed when
```
That information comes...

sboeuf replied:
done

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+
+They both rely on the same model, where they run controller applications
+in the control plane, and inject a proxy as a sidecar inside the service
+pod. The proxy registers to the control plane as a first step, and it
```
The proxy registers in the control plane...

sboeuf replied:
done

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+inside the pod. Those information come from the filtering performed when
+receiving all the traffic initially intended for the service. That's how
+the interaction between the control plane and the proxy allows the user
+to apply load balancing and authentication rules to the on going traffic
```
wouldn't it be better to say something like... to the incoming and outgoing traffic, inside the cluster and between multiple microservices.

sboeuf replied:
done

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+
+I forgot to mention this magic could not happen without a good amount
+of `iptables` rules ensuring the packets reach the proxy instead of
+the expected service. Those rules are setup through an __init__ container
```
rules are set up...

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+
+Follow the [instructions](https://github.com/kata-containers/documentation/blob/master/install/README.md)
+to get Kata Containers properly installed and configured with Kubernetes.
+You can make any choice you want between CRI-O and CRI-containerd, both
```
You can make any choice you want... => You can choose between...

sboeuf replied:
done

raravena80 commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+As mentioned when explaining the basic functioning of those service mesh,
+`iptables` are heavily used, and they need to be properly enabled through
+the guest kernel config, otherwise the __init__ container will not be able
+to perform proper setup of the rules.
```
.. a proper setup...

sboeuf replied:
done

jodh-intel commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+to apply load balancing and authentication rules to the on going traffic
+inside the cluster between multiple micro-services.
+
+I forgot to mention this magic could not happen without a good amount
```
It gave me a good laugh but I'd drop, "I forgot to mention this magic".

sboeuf replied:
oh noooo.... I liked it !
Okay, I'll drop it...

sboeuf replied:
done

jodh-intel commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+I forgot to mention this magic could not happen without a good amount
+of `iptables` rules ensuring the packets reach the proxy instead of
+the expected service. Those rules are setup through an __init__ container
+because they have to be there as soon as the proxy starts.
```
I really think a diagram in this section would help to understand these concepts.

jodh-intel commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+As documented [here](https://github.com/runconduit/conduit/issues/982),
+a kernel version between 4.14.22 and 4.14.40 will cause a deadlock when
+`getsockopt()` get called with `SO_ORIGINAL_DST` option. Unfortunately,
+both service mesh use this system call with this same option from the
```
s/service mesh/service meshes/ ?

jodh-intel commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,229 @@
+
+Here is a summary of what you need to install Istio 0.8.0 on your system
+```
+curl -L https://git.io/getLatestIstio | sh -
````
Please prefix all command lines the user is expected to run with `$ ` as documented in https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md.

sboeuf replied:
done

jodh-intel commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,229 @@
+Here is a summary of what you need to install Istio 0.8.0 on your system
+```
+curl -L https://git.io/getLatestIstio | sh -
+cd istio-0.8.0
````
This line is going to break. How about:

```
$ cd istio-*
```

You could then avoid specifying the version in the text before the code block? If you do this, I'd still recommend documenting the version you used in the commit for reference.

sboeuf replied:
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+
+A service mesh is a powerful way to monitor and control the traffic
+between micro-services running in your Kubernetes cluster. They are
+powerful tools that you might want to use in combination with the
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+# Kata Containers and service mesh for Kubernetes
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+# Kata Containers and service mesh for Kubernetes
+
+A service mesh is a powerful way to monitor and control the traffic
+between micro-services running in your Kubernetes cluster. They are
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+
+## How do they work ?
+
+They both rely on the same model, where they run controller applications
```
"they" relates to both Istio and Conduit. I am going to put both names in here.

And yes you're right, an `Assumptions` section will help.

sboeuf replied:
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,229 @@
+They both rely on the same model, where they run controller applications
+in the control plane, and inject a proxy as a sidecar inside the service
+pod. The proxy registers to the control plane as a first step, and it
+constantly sends different sort of information about the service running
```
done

sboeuf commented:
@raravena80 thanks for the review, I have addressed your comments. Also, regarding `envoy`, I mentioned it line 135 of the doc.

@jodh-intel I have addressed your comments, PTAL !

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+$ export PATH=$PWD/bin:$PATH
+```
+
+Let's now deploy Istio in the control plane of your cluster
````
Now deploy Istio in the control plane of your cluster with the following:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+$ kubectl apply -f install/kubernetes/istio-demo.yaml
+```
+
+In order to verify the control plane has been properly deployed, you
````
To verify that the control plane is properly deployed, you can use both of the following commands:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+### Conduit
+
+As a reference, you can follow Conduit [instructions](https://conduit.io/getting-started/index.html).
```
As a reference, follow the Conduit [instructions](https://conduit.io/getting-started/index.html).

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+As a reference, you can follow Conduit [instructions](https://conduit.io/getting-started/index.html).
+
+Here is a summary of what you need to install Conduit on your system
```
The following is a summary of what you need to install Conduit on your system:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+$ export PATH=$PATH:$HOME/.conduit/bin
+```
+
+Let's now deploy Conduit in the control plane of your cluster
````
Now deploy Conduit in the control plane of your cluster with the following:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+$ conduit install | kubectl apply -f -
+```
+
+In order to verify the control plane has been properly deployed, you
````
To verify that the control plane is properly deployed, you can use both of the following commands:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+## Inject your services with sidecars
+
+Once the control plane is running, you need a deployment defining a
```
Lines 124-127 suggested rewrite: 
Once the control plane is running, you need a deployment to define a few services that rely on each other. Then, you inject the YAML file with the sidecar proxy using the tools provided by each service mesh.

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+file with the sidecar proxy, using the tools provided by each service
+mesh.
+
+If you don't have such a deployment ready, you can refer to the samples
```
If you do not have such a deployment ready, refer to the samples

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+### Istio
+
+Istio provides a [bookinfo](https://istio.io/docs/guides/bookinfo/)
+sample on which you can rely on to inject their `envoy` proxy as a
```
sample, which you can rely on to inject their `envoy` proxy as a

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+sidecar.
+
+You need to use their tool called `istioctl kube-inject` to inject
+your YAML file. We are taking their [bookinfo](https://istio.io/docs/guides/bookinfo/)
```
Lines 139-140 suggested rewrite: ... We use their [bookinfo](https://istio.io/docs/guides/bookinfo/) sample as example:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+### Conduit
+
+Conduit provides an [emojivoto](https://conduit.io/getting-started/index.html)
+sample on which you can rely on to inject their `conduit` proxy as
```
sample, which you can rely on to inject their `conduit` proxy as

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+a sidecar.
+
+You need to use their tool called `conduit inject` to inject your YAML
+file. We are taking their [emojivoto](https://conduit.io/getting-started/index.html)
```
Lines 152-153 suggested rewrite: ... We use their [emojivoto](https://conduit.io/getting-started/index.html) sample as example:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+## Run your services with Kata
+
+Now that you have your service deployment injected with the appropriate
```
Lines 161-162 suggested rewrite: 
Now that your service deployment is injected with the appropriate sidecar containers, manually edit your deployment to make it work with Kata.

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+### Lower privileges
+
+In Kubernetes, the __init__ container is often `privileged` as it needs to
+setup the environment, which often needs some root privileges. But in case
```
Lines 168-170 suggested rewrite: 
... In the case of those services meshes, all they need is the `NET_ADMIN` capability to modify the underlying network rules.

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+they can modify the underlying network rules. Conduit, by default, does not
+use `privileged` container, but Istio does.
+
+For this reason, if you are using Istio, you will need to switch all
```
Because of the previous reason, if you use Istio you need to switch all

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+### Add annotations
+
+This last section is the common part which does not make any difference
```
Lines 178-180 suggested rewrite:
There is no difference between Istio and Conduit in this section. It is about which CRI implementation you use.

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+between Istio and Conduit because it is about which CRI implementation
+you are using.
+
+For both CRI-O and CRI-containerd, you will have to add an annotation
```
For both CRI-O and CRI-containerd, you have to add an annotation

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+__CRI-O__
+
+Add the following annotation for CRI-O
```
Add a colon after CRI-O:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+```yaml
+io.kubernetes.cri-o.TrustedSandbox: "false"
+```
+Here is what your YAML could look like 
````
The following is an example of what your YAML can look like:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+__CRI-containerd__
+
+Add the following annotation for CRI-containerd
```
Add a colon after CRI-containerd:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
````diff
@@ -0,0 +1,251 @@
+```yaml
+io.kubernetes.cri.untrusted-workload: "true"
+```
+Here is what your YAML could look like 
````
The following is an example of what your YAML can look like:

sboeuf replied:
done

klynnrif commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+### Deploy !
+
+You only need to deploy your application at this point, simply using
```
Deploy your application by using the following:

sboeuf replied:
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+## Assumptions
+
+You are expected to be familiar with concepts such as __pods__,
+__containers__, __control plane__, __data plane__ and __sidecar__.
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+a first step, and it constantly sends different sorts of information about
+the service running inside the pod. That information comes from the
+filtering performed when receiving all the traffic initially intended for
+the service. That's how the interaction between the control plane and the
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+filtering performed when receiving all the traffic initially intended for
+the service. That's how the interaction between the control plane and the
+proxy allows the user to apply load balancing and authentication rules to
+the incoming and outgoing traffic, inside the cluster and between multiple
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+the incoming and outgoing traffic, inside the cluster and between multiple
+micro-services.
+
+This could not happen without a good amount of `iptables` rules ensuring
```
done

sboeuf commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,251 @@
+
+Follow the [instructions](https://github.com/kata-containers/documentation/blob/master/install/README.md)
+to get Kata Containers properly installed and configured with Kubernetes.
+You can choose between CRI-O and CRI-containerd, both will be supported
```
done

sboeuf commented:
Thanks @klynnrif for the thorough review ! Comments addressed, PTAL.

jodh-intel commented on svc-mesh/service-mesh.md:
```diff
@@ -0,0 +1,248 @@
+# Kata Containers and service mesh for Kubernetes
+
+* [Assumptions](#assumptions)
+* [ How do they work ?](#how-do-they-work-)
```
Micro nit: extraneous space before "How".

sboeuf replied:
done
